src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,222 @@
+
+    private int size = 0;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(value);
+            tail = head;
+        } else {
+            Node insertedElement = new Node(value);
+            insertedElement.prev = tail;
+            tail.next = insertedElement;
+            tail = insertedElement;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        verifyInBound(index);
+        if (index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            Node<T> insertedElement = new Node<>(value);
+            insertedElement.next = head;
+            head.prev = insertedElement;
+            head = insertedElement;
+            size++;
+            return;
+        }
+
+        Node<T> insertedBeforeNode = find(index);
+        Node<T> insertedElement = new Node<>(value);
+
+        insertedBeforeNode.prev.next = insertedElement;
+        insertedElement.prev = insertedBeforeNode.prev;
+        insertedElement.next = insertedBeforeNode;
+        insertedBeforeNode.prev = insertedElement;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isEmpty() || index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("incorrect index: " + index);
+        } else {
+            return (T) find(index).value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (isEmpty() || index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("incorrect index: " + index);
+        } else if (index == 0) {
+            Node<T> insertedElement = new Node<>(value);
+            insertedElement.next = head.next;
+
+            T oldValue = head.value;
+
+            head.next.prev = insertedElement;
+            head = insertedElement;
+            return oldValue;
+        } else if (index == size - 1) {
+            Node<T> insertedElement = new Node<>(value);
+            insertedElement.prev = tail.prev;
+
+            T oldValue = tail.value;
+
+            tail.prev.next = insertedElement;
+            tail = insertedElement;
+            return oldValue;
+        } else {
+            Node<T> oldElement = head;
+            for (int i = 0; i < index; i++) {
+                oldElement = oldElement.next;
+            }
+            Node<T> insertedElement = new Node<>(value);
+            insertedElement.prev = oldElement.prev;
+            insertedElement.next = oldElement.next;
+            oldElement.prev.next = insertedElement;
+            oldElement.next.prev = insertedElement;
+            return oldElement.value;
+        }
```

Comment:
you dont need to create a new node in this method - just change the value of a node